packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart

```diff
@@ -224,23 +254,80 @@ abstract class RenderSliverMultiBoxAdaptor extends RenderSliver
 
   bool _debugAssertChildListLocked() => childManager.debugAssertChildListLocked();
 
+  /// Verify that the child list index is in strictly increasing order.
+  ///
+  /// This has no effect in release builds..
+  void _debugVerifyChildOrder(){
+    assert(() {
+      if (_debugChildIntegrityEnabled) {
+        if (firstChild != null) {
+          int index = indexOf(firstChild);
+          RenderBox child = childAfter(firstChild);
+          while (child != null) {
+            assert(indexOf(child) > index);
+            index = indexOf(child);
+            child = childAfter(child);
+          }
+        }
+      }
+      return true;
+    }());
+  }
+
   @override
   void insert(RenderBox child, { RenderBox after }) {
     assert(!_keepAliveBucket.containsValue(child));
     super.insert(child, after: after);
     assert(firstChild != null);
     assert(() {
-      int index = indexOf(firstChild);
-      RenderBox child = childAfter(firstChild);
-      while (child != null) {
-        assert(indexOf(child) > index);
-        index = indexOf(child);
-        child = childAfter(child);
-      }
+      _debugVerifyChildOrder();
       return true;
     }());
   }
 
+  @override
+  void move(RenderBox child, { RenderBox after }) {
+    // There are two scenarios:
+    //
+    // 1. The child is not keptAlive.
+    // The child is in childList maintained by ContainerRenderObjectMixin.
+    // We can call the super.move and update parentData with new slot.
+    //
+    // 2. The child is keptAlive.
+    // In this case, the child is no longer in the childList but might be stored in
+    // [_keepAliveBucket]. We need to update the location of the child in the bucket.
+    final SliverMultiBoxAdaptorParentData childParentData = child.parentData;
+    if (!childParentData.keptAlive) {
+      super.move(child, after: after);
+      // The following call updates the slot in the parentData.
+      childManager.didAdoptChild(child);
```

Comment:
please update the documentation for `didAdoptChild` to mention this new contract.

Reply:
I will update the first sentence, but I think the original documentation is good enough?

  /// Subclasses must ensure that the [SliverMultiBoxAdaptorParentData.index]
  /// field of the child's [RenderObject.parentData] accurately reflects the
  /// child's index in the child list after this function returns.